### PR TITLE
PERF: DataFrame.transpose for pyarrow-backed

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v2.1.0.rst` file if fixing a bug or adding a new feature.

Similar to #52836, but for pyarrow-backed frames. 

```
import pandas as pd
import pyarrow as pa
import numpy as np

values = np.random.randn(1000, 1000)
df = pd.DataFrame(values, dtype=pd.ArrowDtype(pa.float64()))

%timeit df.T

# 346 ms ± 67.5 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    -> main
# 42.5 ms ± 2.12 ms per loop (mean ± std. dev. of 7 runs, 10 loops each) -> PR
```

Renamed the masked implementation from `homogenous` -> `homogeneous`. Both may be valid(?), but `homogeneous` is used elsewhere.

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -3648,17 +3648,34 @@ def transpose(self, *args, copy: bool = False) -> DataFrame:
...
+                # We have arrow EAs with the same dtype. We can transpose faster.
+                from pandas.core.arrays.arrow.array import (
+                    ArrowExtensionArray,
+                    transpose_homogeneous_arrow_extension_arrays,
+                )
+
+                if isinstance(self._mgr, ArrayManager):
+                    arrays = self._mgr.arrays
+                else:
+                    arrays = list(self._iter_column_arrays())
```
Comment (jbrockmendel) [inline pandas/core/frame.py:3675]: cant do this unconditionally?

Comment (lukemanley) [reply pandas/core/frame.py:3675]: Do you mean use `_iter_column_arrays` in both cases? The masked array implementation just above this uses the condition as well, presumably for performance:

```
import pandas as pd

pd.options.mode.data_manager = "array"

df = pd.DataFrame(columns=range(100000), dtype="int64[pyarrow]")

%timeit list(df._iter_column_arrays())
# 57.9 ms ± 2.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit df._mgr.arrays
# 110 ns ± 8.79 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Comment (jbrockmendel) [reply pandas/core/frame.py:3675]: Fair enough, thanks for taking a look

Comment (lukemanley) [reply pandas/core/frame.py]: on second thought, I moved the condition into `DataFrame._iter_column_arrays` to avoid repeating it elsewhere

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @lukemanley

**Pull request merged**